sdk/go/pulumix/compose.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pulumix
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"runtime"
+	"runtime/debug"
+
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	"github.com/pulumi/pulumi/sdk/v3/go/internal"
+)
+
+// Composer allows composing multiple outputs together in a type-safe manner.
+//
+// To use Composer, begin a composition with the [Compose] function.
+// This returns an Output[O] that will be fulfilled with the value returned by f.
+//
+//	o := pulumix.Compose(ctx, func(c *pulumix.Composer) (O, error) {
+//		// ...
+//	})
+//
+// Inside the function, use [ComposeAwait] to await other outputs.
+//
+//	v1 := pulumix.ComposeAwait(c, o1)
+//	v2 := pulumix.ComposeAwait(c, o2)
+//	// ...
+//
+// Combine the values into a single result and return it.
+// For example:
+//
+//	return &MyStruct{
+//		Foo: v1,
+//		Bar: v2,
+//		...
+//	}, nil
+//
+// If f returns an error, the output will be rejected.
+// If any of the awaited outputs fail, the composition will abort
+// and the output will be rejected.
+//
+// Composer allows combining an arbitrary number of outputs.
+// For simpler cases, consider [Apply] or any of the ApplyN variants.
```

Comment:
Perhaps a warning along the lines of "do not capture *pulumix.Composer outside of the scope of the function". Or it's obvious? This wouldn't work:

```go
var composer *pulumix.Composer

pulumix.Compose(ctx, func(c *pulumix.Composer) (O, error) {
   composer = c
})

// Yay I can await outputs now!
v1 := pulumix.ComposeAwait(composer, o1)
```

Reply:
There's a warning:

https://github.com/pulumi/pulumi/pull/13829/files/68405d842f44d902630dfb632aea136ac56b0000#diff-7174e485d1e36d95f310cedd29e0c28193afa72176c7da2dc48c20f536d0ed33R81

There's also a test for it. :)